x/mm2/keeper/msg_server_test.go

```diff
@@ -0,0 +1,218 @@
+	s.Require().NoError(err)
+	s.Require().NotNil(resp)
+
+	// query the market map
+	mm, err := s.keeper.GetAllMarketsMap(s.ctx)
+	s.Require().NoError(err)
+	s.Require().Equal(types.MarketMap{Markets: mm}, types.MarketMap{
+		Markets: map[string]types.Market{
+			btcusdt.Ticker.String(): btcusdt,
+			usdtusd.Ticker.String(): usdtusd,
+		},
+	})
+
+	// TODO: test hooks
+
+	s.Run("unable to process for invalid authority", func() {
+		msg = &types.MsgCreateMarkets{
+			Authority: "invalid",
+		}
+		resp, err = msgServer.CreateMarkets(s.ctx, msg)
+		s.Require().Error(err)
+		s.Require().Nil(resp)
+	})
+
+	// set a market in the map
+	s.Run("unable to process nil request", func() {
+		resp, err = msgServer.CreateMarkets(s.ctx, nil)
+		s.Require().Error(err)
+		s.Require().Nil(resp)
+	})
+
+	s.Run("unable to create market that already exists", func() {
+		msg = &types.MsgCreateMarkets{
+			Authority: s.authority.String(),
+			CreateMarkets: []types.Market{
+				btcusdt,
+			},
+		}
+		resp, err = msgServer.CreateMarkets(s.ctx, msg)
+		s.Require().Error(err)
+		s.Require().Nil(resp)
+	})
+
+	s.Run("unable to create market with normalize by that is not on chain tickers", func() {
+		msg = &types.MsgCreateMarkets{
+			Authority: s.authority.String(),
+			CreateMarkets: []types.Market{
+				{
+					Ticker: ethusdt.Ticker,
+					ProviderConfigs: []types.ProviderConfig{
+						{
+							Name:           "kucoin",
+							OffChainTicker: "eth-usdt",
+							NormalizeByPair: &slinkytypes.CurrencyPair{
+								Base:  "INVALID",
+								Quote: "PAIR",
+							},
+						},
+					},
+				},
+			},
+		}
+		resp, err = msgServer.CreateMarkets(s.ctx, msg)
+		s.Require().Error(err)
+		s.Require().Nil(resp)
+	})
+}
+
+func (s *KeeperTestSuite) TestMsgServerUpdateMarkets() {
+	msgServer := keeper.NewMsgServer(s.keeper)
+
+	// create initial markets
+	createMsg := &types.MsgCreateMarkets{
+		Authority: s.authority.String(),
+		CreateMarkets: []types.Market{
+			btcusdt,
+			usdtusd,
+		},
+	}
+	createResp, err := msgServer.CreateMarkets(s.ctx, createMsg)
+	s.Require().NoError(err)
+	s.Require().NotNil(createResp)
+
+	// query the market map
+	mm, err := s.keeper.GetAllMarketsMap(s.ctx)
+	s.Require().NoError(err)
+	s.Require().Equal(types.MarketMap{Markets: mm}, types.MarketMap{
+		Markets: map[string]types.Market{
+			btcusdt.Ticker.String(): btcusdt,
+			usdtusd.Ticker.String(): usdtusd,
+		},
+	})
+
+	// TODO: test hooks
+
+	s.Run("unable to process for invalid authority", func() {
+		msg := &types.MsgUpdateMarkets{
+			Authority: "invalid",
+		}
+		resp, err := msgServer.UpdateMarkets(s.ctx, msg)
+		s.Require().Error(err)
+		s.Require().Nil(resp)
+	})
+
+	// set a market in the map
+	s.Run("unable to process nil request", func() {
+		resp, err := msgServer.UpdateMarkets(s.ctx, nil)
+		s.Require().Error(err)
+		s.Require().Nil(resp)
+	})
+
+	s.Run("able to update market that already exists", func() {
+		tickerUpdate := btcusdt
+		tickerUpdate.Ticker.Decimals = 1
```

Comment:
I would bring this up to the team, but do we anticipate that there are certain things that should not be able to change after the market is initially created?

Reply:
The only one that I can really think of is that the currency pair does not change

Reply:
which is already enforced